### PR TITLE
AuthN: Add service account token generation to `ExtSvcAccountsService`

### PR DESCRIPTION
**Pull request by gamab**
**What is this feature?**

This PR adds to the `ExtSvcAccountsService` service, code to generate and store service account tokens. `ExtSvcAccountsService` also now implements the `ExternalServiceRegistry` interface allowing it to be used as an `authProvider` in a later PR.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

https://github.com/grafana/grafana-enterprise/issues/5824

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:56]: This is to implement the `extsvcauth.ExternalServiceRegistry` interface.
We will use this in another PR to allow plugins to request service account tokens.

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:32]: I'm not entirely sure I'm allowed to us the SQL store straight away. Need to double check with the plugin squad :slightly_smiling_face:

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:218]: Eventually all these Credentials related methods could be used by the `OAuth2Server` as well to have an HA compliant credentials management.

**Review by Jguer — APPROVED** [review]

File: pkg/services/extsvcauth/extsvcaccounts/models.go
Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/models.go:9]: It was for "secret key value" Type. Any suggestion?

File: pkg/services/extsvcauth/extsvcaccounts/models.go
Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/models.go:13]: Very good point. Will do now

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:179]: Eventually all these `Credentials` related methods could be used by the `OAuth2Server` as well to have an HA compliant credentials management.

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (andresmgot) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:225]: Errors generated withi errorutil are `Base` errors that are meant to be extended with a real error. You can either do here something like:

```go
return nil, extsvcauth.ErrCredentialsNotFound.Errorf("credentials not found")
```

Or simply export the result of calling Errorf as `ErrCredentialsNotFound`

File: pkg/services/extsvcauth/extsvcaccounts/models.go
Comment (andresmgot) [reply pkg/services/extsvcauth/extsvcaccounts/models.go:9]: maybe `secretKVStoreType`?

Comment (kalleep) [reply pkg/services/extsvcauth/extsvcaccounts/models.go:9]: seems to be used as a prefix for names, so namePrefix?

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (andresmgot) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:62]: should this at least log as an error?

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (andresmgot) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:78]: isn't this an error?

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (andresmgot) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:191]: if this is logged every time a token is retrieved maybe should be `Debug`

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (andresmgot) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:32]: does not sound wrong to me but better double check with the @grafana/backend-platform who owns that.

**Review by andresmgot — COMMENTED** [review]
Some minor comments here. In general it looks good to me :+1:

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:225]: Thanks, it's a miss on my side

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:62]: I think `Warning` is fine there, we can still do the service account setup and generate a token.
I guess it depends on: Would you error out here?

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:78]: The only case you can get there is if the registration has: `Self.Enabled = false`, no permission, or no registration has been provided at all. 
In those cases, it's expected not to setup any token.

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:191]: It's logged only the first time the token is created :+1:. Afterwards, the token being retrieved from the secret store, we won't re-generate it.

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (mildwonkey) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:32]: This looks perfectly reasonable to me! We've certainly used this pattern before in places where we could not work around dependency cycles; sometimes it's the only pragmatic option. That's especially true in this case where the kvstore service directly calls the underlying store for all its methods with no extra business logic.

Comment (marefr) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:32]: Even though codeowner is grafana backend think @grafana/grafana-operator-experience-squad own the unified secrets feature.

Would it help if `kvstore.SecretsKVStore` interface was defined outside of the kvstore package and then you could let wire provide the `SecretsKVStore` dependency?

Comment (marefr) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:32]: One potential problem with this might be that you take a shortcut around the kvstore and its cache and in addition secrets manager plugin would not be supported if not mistaken

Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:32]: > Would it help if `kvstore.SecretsKVStore` interface was defined outside of the kvstore package and then you could let wire provide the `SecretsKVStore` dependency?

Not sure I follow. :slightly_smiling_face: If I understood correctly, the problem lies in the fact that the `SecretKVStore` interface could be provided by a grpc plugin, and since this service aims to set-up credentials for plugins, we have a cyclic dependency injection. 

> One potential problem with this might be that you take a shortcut around the kvstore and its cache

Cache wise, I think this is fine, since this reading/saving is done once on startup, the cache is not necessary.

> in addition secrets manager plugin would not be supported if not mistaken

This is the part I was more concerned about: circumventing the secrets manager plugin.

Comment (marefr) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:32]: > Edit: Ah I guess there is a confusion 🙂 I'm talking about a cyclic dependency injection problem not an cyclic import problem.

Ohhh I see sorry :cry: Now I remember we've had this problem earlier. It's definitely something we need to figure out moving forward if plugins are supposed to provide core functionality. I'm going to discuss it with my team. 

I guess @grafana/grafana-operator-experience-squad has the verdict if this is a viable way forward for now

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [inline pkg/services/extsvcauth/extsvcaccounts/service.go:217]: Eventually all these `Credentials` related methods could be used by the `OAuth2Server` as well to have an HA compliant credentials management.

File: pkg/services/extsvcauth/extsvcaccounts/service.go
Comment (gamab) [reply pkg/services/extsvcauth/extsvcaccounts/service.go:32]: I got a green light from https://github.com/orgs/grafana/teams/grafana-operator-experience-squad :heavy_check_mark:

**Comment by gamab** [comment]
Thanks everyone for the reviews :heart:

**Pull request merged**
